federation-2/router-bridge/js-src/plan.ts

```diff
@@ -21,63 +47,463 @@ export class BridgeQueryPlanner {
+// between queries that are similar to each other, and if for some reason your
+// GraphQL client generates query strings with elements in nondeterministic
+// order, it can make sure the queries are treated as identical.
+function sortAST(ast: DocumentNode): DocumentNode {
+  return visit(ast, {
+    Document(node: DocumentNode) {
+      return {
+        ...node,
+        // Use sortBy instead of sorted because 'definitions' is not optional.
+        // The sort on "kind" places fragments before operations within the document
+        definitions: sortBy(node.definitions, "kind", "name.value"),
+      };
+    },
+    OperationDefinition(
+      node: OperationDefinitionNode
+    ): OperationDefinitionNode {
+      return {
+        ...node,
+        variableDefinitions: sorted(
+          node.variableDefinitions,
+          "variable.name.value"
+        ),
+      };
+    },
+    SelectionSet(node: SelectionSetNode): SelectionSetNode {
+      return {
+        ...node,
+        // Define an ordering for field names in a SelectionSet.  Field first,
+        // then FragmentSpread, then InlineFragment.  By a lovely coincidence,
+        // the order we want them to appear in is alphabetical by node.kind.
+        // Use sortBy instead of sorted because 'selections' is not optional.
+        selections: sortBy(node.selections, "kind", "name.value"),
+      };
+    },
+    Field(node: FieldNode): FieldNode {
+      return {
+        ...node,
+        arguments: sorted(node.arguments, "name.value"),
+      };
+    },
+    FragmentSpread(node: FragmentSpreadNode): FragmentSpreadNode {
+      return { ...node, directives: sorted(node.directives, "name.value") };
+    },
+    InlineFragment(node: InlineFragmentNode): InlineFragmentNode {
+      return { ...node, directives: sorted(node.directives, "name.value") };
+    },
+    FragmentDefinition(node: FragmentDefinitionNode): FragmentDefinitionNode {
+      return {
+        ...node,
+        directives: sorted(node.directives, "name.value"),
+        variableDefinitions: sorted(
+          node.variableDefinitions,
+          "variable.name.value"
+        ),
+      };
+    },
+    Directive(node: DirectiveNode): DirectiveNode {
+      return { ...node, arguments: sorted(node.arguments, "name.value") };
+    },
+  });
+}
+
+// Like lodash's sortBy, but sorted(undefined) === undefined rather than []. It
+// is a stable non-in-place sort.
+function sorted<T>(
+  items: ReadonlyArray<T> | undefined,
+  ...iteratees: string[]
+): T[] | undefined {
+  if (items) {
+    return sortBy(items, ...iteratees);
+  }
+  return undefined;
+}
+
+// removeAliases gets rid of GraphQL aliases, a feature by which you can tell a
+// server to return a field's data under a different name from the field
+// name. Maybe this is useful if somebody somewhere inserts random aliases into
+// their queries.
+function removeAliases(ast: DocumentNode): DocumentNode {
```

Comment:
Aliases are also used if we want to return the same field multiple times but with different directives, should they be removed in that case?

Reply:
This code has been duplicated from the gateways to make sure the behavior remains the same. I agree it's however worth discussing once we refactor (or riir?) it